File: core/src/main/java/fr/sncf/osrd/api/pathfinding/constraints/ElectrificationConstraints.java
package fr.sncf.osrd.api.pathfinding.constraints;

import static fr.sncf.osrd.api.utils.PathPropUtils.makePathProps;

import com.google.common.collect.Range;
import com.google.common.collect.RangeSet;
import com.google.common.collect.TreeRangeSet;
import fr.sncf.osrd.sim_infra.api.BlockInfra;
import fr.sncf.osrd.sim_infra.api.PathProperties;
import fr.sncf.osrd.sim_infra.api.RawSignalingInfra;
import fr.sncf.osrd.train.RollingStock;
import fr.sncf.osrd.utils.DistanceRangeMap;
import fr.sncf.osrd.utils.graph.Pathfinding;
import fr.sncf.osrd.utils.graph.functional_interfaces.EdgeToRanges;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

public record ElectrificationConstraints(
        BlockInfra blockInfra,
        RawSignalingInfra rawInfra,
        Collection<RollingStock> rollingStocks
) implements EdgeToRanges<Integer> {

    @Override
    public Collection<Pathfinding.Range> apply(Integer blockId) {
        var res = new HashSet<Pathfinding.Range>();
        var path = makePathProps(blockInfra, rawInfra, blockId);
        for (var stock : rollingStocks)
            res.addAll(getBlockedRanges(stock, path));
        return res;
    }

    /**
     * Returns the sections of the given block that can't be used by the given rolling stock
     * because it needs electrified tracks and isn't compatible with the catenaries in some range
     */
    private static Set<Pathfinding.Range> getBlockedRanges(RollingStock stock, PathProperties path) {
        if (!stock.isElectricOnly())
            return Set.of();

        var res = new HashSet<Pathfinding.Range>();
        var voltages = path.getCatenary();
        var neutralSections = rangeSetFromMap(path.getNeutralSections());
        for (var section : voltages) {
            if (section.getLower() == section.getUpper())
                continue;
            if (!stock.getModeNames().contains(section.getValue())) {
                var voltageInterval = Range.open(section.getLower(), section.getUpper());
                var blockingRanges = neutralSections.complement().subRangeSet(voltageInterval).asRanges();

                for (var blockingRange : blockingRanges) {
                    assert blockingRange.lowerEndpoint() < blockingRange.upperEndpoint();
                    res.add(new Pathfinding.Range(blockingRange.lowerEndpoint(), blockingRange.upperEndpoint()));
                }
            }
        }
        return res;
    }

    private static <T> RangeSet<Long> rangeSetFromMap(DistanceRangeMap<T> rangeMap) {
        return TreeRangeSet.create(
                rangeMap.asList().stream()
                        .map(entry -> Range.closed(entry.getLower(), entry.getUpper()))
                        .collect(Collectors.toSet())
        );
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/constraints/ElectrificationConstraints.kt
package fr.sncf.osrd.api.pathfinding.constraints

import com.google.common.collect.Range
import com.google.common.collect.RangeSet
import com.google.common.collect.TreeRangeSet
import fr.sncf.osrd.api.pathfinding.makePathProps
import fr.sncf.osrd.graph.Pathfinding
import fr.sncf.osrd.graph.PathfindingConstraint
import fr.sncf.osrd.sim_infra.api.Block
import fr.sncf.osrd.sim_infra.api.BlockId
import fr.sncf.osrd.sim_infra.api.BlockInfra
import fr.sncf.osrd.sim_infra.api.PathProperties
import fr.sncf.osrd.sim_infra.api.RawSignalingInfra
import fr.sncf.osrd.utils.DistanceRangeMap
import fr.sncf.osrd.utils.units.Distance
import fr.sncf.osrd.utils.units.Offset
import java.util.stream.Collectors

data class ElectrificationConstraints(
    val blockInfra: BlockInfra,
    val rawInfra: RawSignalingInfra,
    val compatibleElectrification: Collection<String>
) : PathfindingConstraint<Block> {
    override fun apply(edge: BlockId): Collection<Pathfinding.Range<Block>> {
        val path = makePathProps(blockInfra, rawInfra, edge)
        return getBlockedRanges(path, compatibleElectrification)
    }

    companion object {
        /**
         * Returns the sections of the given block that can't be used by the given rolling stock
         * because it needs electrified tracks and isn't compatible with the electrifications in
         * some range
         */
        private fun getBlockedRanges(
            path: PathProperties,
            compatibleElectrification: Collection<String>
        ): Set<Pathfinding.Range<Block>> {
            val res = HashSet<Pathfinding.Range<Block>>()
            val voltages = path.getElectrification()
            val neutralSections = rangeSetFromMap(path.getNeutralSections())
            for ((lower, upper, value) in voltages) {
                if (lower == upper) continue
                if (!compatibleElectrification.contains(value)) {
                    val voltageInterval = Range.open(lower, upper)
                    val blockingRanges =
                        neutralSections.complement().subRangeSet(voltageInterval).asRanges()
                    for (blockingRange in blockingRanges) {
                        assert(blockingRange.lowerEndpoint() < blockingRange.upperEndpoint())
                        res.add(
                            Pathfinding.Range(
                                Offset(blockingRange.lowerEndpoint()),
                                Offset(blockingRange.upperEndpoint())
                            )
                        )
                    }
                }
            }
            return res
        }

        private fun <T> rangeSetFromMap(rangeMap: DistanceRangeMap<T>): RangeSet<Distance> {
            return TreeRangeSet.create(
                rangeMap
                    .asList()
                    .stream()
                    .map { (lower, upper): DistanceRangeMap.RangeMapEntry<T> ->
                        Range.closed(lower, upper)
                    }
                    .collect(Collectors.toSet())
            )
        }
    }
}
